### PR TITLE
[Merged by Bors] - feat(Data/Set/Function): add `BijOn.insert` and `BijOn.sdiff_singleton`

### PR DESCRIPTION
**Pull request by pitmonticone**
Co-authored-by: Daniel Weber

Upstreamed from the [EquationalTheories](https://github.com/teorth/equational_theories) project.

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Data/Set/Function.lean [hunk 1]
```diff
@@ -947,6 +948,19 @@ theorem BijOn.subset_left {r : Set α} (hf : BijOn f s t) (hrs : r ⊆ s) :
     BijOn f r (f '' r) :=
   (hf.injOn.mono hrs).bijOn_image
 
+theorem BijOn.insert (h₁ : BijOn f s t) (h₂ : f a ∉ t) :
+    BijOn f (insert a s) (insert (f a) t) := by
```
Comment (eric-wieser) [inline Mathlib/Data/Set/Function.lean:952]: Is there an `Iff` version of this?

**Comment by YaelDillies** [comment]
Please avoid pinging people in the PR description. They will thank you later

**Comment by pitmonticone** [comment]
Ops, you're right.

I should have added "Author Name <email>".

Sorry for that!

File: Mathlib/Data/Set/Function.lean (hunk 1 above)
Comment (Command-Master) [reply Mathlib/Data/Set/Function.lean:952]: The other direction also needs `a ∉ s`, if I'm not mistaken

**Comment by YaelDillies** [comment]
I've fixed it now. It only becomes a problem once the PR is merged, as then the ping makes it to the commit message.

**Comment by pitmonticone** [comment]
Thank you, Yaël.

File: Mathlib/Data/Set/Function.lean [hunk 2]
```diff
@@ -947,6 +948,19 @@ theorem BijOn.subset_left {r : Set α} (hf : BijOn f s t) (hrs : r ⊆ s) :
...
+  repeat rw [insert_eq]
+  refine Set.BijOn.union (bijOn_singleton.mpr rfl) h₁ ?_
```
Comment (YaelDillies) [inline Mathlib/Data/Set/Function.lean:954]:
```suggestion
  refine (bijOn_singleton.mpr rfl).union h₁ ?_
```

File: Mathlib/Data/Set/Function.lean (hunk 1 above)
Comment (YaelDillies) [reply Mathlib/Data/Set/Function.lean:952]: Worth having as well, I would say

File: Mathlib/Data/Set/Function.lean [hunk 3]
```diff
@@ -292,7 +292,8 @@ theorem mapsTo_singleton {x : α} : MapsTo f {x} t ↔ f x ∈ t :=
 theorem mapsTo_empty (f : α → β) (t : Set β) : MapsTo f ∅ t :=
   empty_subset _
 
-@[simp] theorem mapsTo_empty_iff : MapsTo f s ∅ ↔ s = ∅ := by
+@[simp]
+theorem mapsTo_empty_iff : MapsTo f s ∅ ↔ s = ∅ := by
```
[the remark this answers is not in the document]
Comment (pitmonticone) [reply Mathlib/Data/Set/Function.lean:296]: Ops, there is no point. I'll revert later.

**Comment by YaelDillies** [comment]
Can you add the iffs as per https://github.com/leanprover-community/mathlib4/pull/17971#discussion_r1810296080 ?

**Comment by pitmonticone** [comment]
Ok, I'll add it later today or tomorrow. Thank you for your review.

**Comment by YaelDillies** [comment]
Scooped 😁 Think it all looks good now. Thanks Daniel

maintainer merge

**Pull request closed without merging**